File: LeetCode/Java/MinimumIndexSumofTwoLists.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MinimumIndexSumofTwoLists {

    public String[] findRestaurant(String[] list1, String[] list2) {

        Map<String, Integer> map = new HashMap<>();

        for (int i = 0; i < list1.length; i++) {
            map.put(list1[i], i);
        }

        int minSum = Integer.MAX_VALUE;
        List<String> result = new ArrayList<>();

        for (int j = 0; j < list2.length; j++) {
            if (map.containsKey(list2[j])) {
                int sum = j + map.get(list2[j]);

                if (sum < minSum) {
                    result.clear();
                    result.add(list2[j]);
                    minSum = sum;
                } 
                else if (sum == minSum) {
                    result.add(list2[j]);
                }
            }
        }

        return result.toArray(new String[0]);
    }

    public static void main(String[] args) {
        MinimumIndexSumofTwoLists s = new MinimumIndexSumofTwoLists();

        String[] list1 = {"Shogun", "Tapioca Express", "Burger King", "KFC"};
        String[] list2 = {"Piatti", "The Grill at Torrey Pines", "Hungry Hunter Steakhouse", "Shogun"};

        System.out.println(Arrays.toString(s.findRestaurant(list1, list2)));
    }
}

File: LeetCode/Python/Minimum Index Sum of Two Lists.py
from typing import List

class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        dict1={}
        dict2={}
        for i in list1:
            dict1[i]=list1.index(i)
        for j in list2:
            dict2[j]=list2.index(j)
        merged= {key: dict1.get(key, 0) + dict2.get(key, 0) for key in set(dict1) if key in dict2}
        min_value = min(merged.values())
        mini = [key for key, value in merged.items() if value == min_value]
        return mini
    
s= Solution()

list1 = ["Shogun","Tapioca Express","Burger King","KFC"]
list2 = ["Piatti","The Grill at Torrey Pines","Hungry Hunter Steakhouse","Shogun"]

print(s.findRestaurant(list1,list2))